### ISSUE TITLE
how to call the SharedDatabase class by passing QueryExecutor instance?

### ISSUE THREAD

**Issue by techttiru** [simolus3/drift#1873]
I stuck in how to call a SharedDatabase class by passing QueryExecutor instance. in the below link, i follow all the instructions and it works fine as expected if i run web or mobile separately. but if i want to run both, then i get error. this document explains how to use the shared database between web and native. i am able to follow what has been explained in it. but i dont know how call the Shareddatabase class instance because that is where i have read and write operation methods as it asks the queryexecutor instance.

https://drift.simonbinder.eu/web/

i have followed the exact procedures that is explained in this document. i have created database.dart, native.dart, web.dart, unsupported.dart, shared.dart. Everything is fine. now i need to instantiate SharedDatabase class in statement management to create singleton object and also i need SharedDatabase instance to call insert and read methods that i have written inside SharedDatabase class. now how to instantiate SharedDatabase class as its asking QueryExecutor. what QueryExecutor i should pass here? Thanks.

@DriftDatabase(tables: [Modes])
class SharedDatabase extends _$MyDatabase {
  SharedDatabase(QueryExecutor e) : super(e);

  @override
  int get schemaVersion => 1;

  Future<void> insertNewrowIntoModeTable() async {
    final v1 = ModesCompanion(
        userName: const Value('test  new'),
        watchMode: const Value('f1'),
        modifiedDate: Value(
          DateTime.now(),
        ));

    final v2 = await into(modes).insert(v1);
    print('inserted successfully in db - $v2');
  }

  Future<void> readAllModes() async {
    final f2 = await select(modes).get();
    f2.forEach((element) => print(element.toString()));
  }
}

**Comment by techttiru** [comment]
hey. Thank for the quick response!. Really appreciated!.. i will try this and update to this group!.

**Comment by techttiru** [comment]
@simolus3  Hi, I see this has been closed but i am adding more information on the same thread so that it will be in continues conversation for others easily relate it. as you suggested i am able to access using constructDb(). this is all good and thanks for this package.

additionally i am trying to protect this db with password encryption in flutter web application. in one of the thread i saw your response and i followed your response to add password as below but i am not seeing any effect for this. because i am able to see and open up the db in web browser -> application -> local storage -> web url -> moor_db_str_db.  is there a way we can protect this data? 

```
import 'package:drift/web.dart';
import 'person_table.dart';

SharedDatabase constructDb() {
  return SharedDatabase(WebDatabase('db', logStatements: true, setup: (db) {
    db.run("pragma KEY = 'password';");
  }));
}
```

Also, additionally is it possible to store this sqlite db in the client system instead of storing into client web browser?

Appreciate your response.. please let me know if you need more details or you want me to open up a new issue.

**Comment by simolus3** [comment]
>  is there a way we can protect this data? is the feature still not available for us use?

A `WebDatabase` currently wraps the sql.js library, which doesn't support the `key` pragma out of the box. You could follow the advice given in https://github.com/sql-js/sql.js/issues/54 to obtain a variant of that library (or rather, the wasm file) with support for the `KEY` pragma.

(Also, note that you need `sqlcipher_flutter_libs` instead of `sqlite3_flutter_libs` for the key pragma to work in iOS and Android)

> Also, additionally is it possible to store this sqlite db in the client system instead of storing into client web browser?

Not really, web applications don't have access to that. There's a file system web API, but it only works in Chrome and is in the process of being deprecated everywhere else. And even that would be a hidden, sandboxed file that the user doesn't have direct access to.

**Comment by techttiru** [comment]
@simolus3 . Hey Thanks again for quick response. I looked at the sql.js link that you provided and my understanding is we need to use sqleet instead of sqlite3 or it says  `simply compile sql.js after replacingc/sqlite3.c and c/sqlite3.h with sqleet.c and sqleet.h amalgamations.` but here my understanding is, since i use drift package, drift package is creating sql.js internally when we work on the drift syntax. so i am not able to follow how can i compile the sql.js with the instruction that was provided. can you please give me a example how to follow the given instruction? i am unable to connect the dots here.

Regarding the below comment,
`(Also, note that you need sqlcipher_flutter_libs instead of sqlite3_flutter_libs for the key pragma to work in iOS and Android)`

in my case, my app is for flutter web, android and ios. my primary focus is for flutter web. so sqlcipher_flutter_libs will work for flutter web as well? also my app user always uses chrome. so browser is not the problem in my case. can you please let me know how to store the db in file system in client machine and protect it as needed and access via flutter web application?

**Comment by simolus3** [comment]
> in my case, my app is for flutter web, android and ios. my primary focus is for flutter web. so sqlcipher_flutter_libs will work for flutter web as well?

No, I should have been clearer here. Drift is responsible for generating SQL statements, but it needs additional packages to actually run those statements.

- For native applications (Android, iOS, macOS, Linux, Windows), the `sqlcipher_flutter_libs` package makes sure that SQLCipher is compiled into your application and available when it runs. The `sqlite3` package is then used by drift internally to send those statements. These packages have no effect on the web.
- With a `WebDatabase`, drift uses the sql.js library to run its statements. This is why you had to add `<script defer src="sql-wasm.js"></script>` to your `index.html` to use drift on the web. You can compile sqleet to work with sql.js (but I can't help you with that) and then use that custom version of sql.js with drift.

So drift generates SQL statements, but not the implementation actually running those.

**Comment by simolus3** [comment]
> how to store the db in file system in client machine and protect it as needed and access via flutter web application?

I'm not that familiar with the web system APIs, so I can't give you a complete example here. But you can write a custom [DriftWebStorage](https://pub.dev/documentation/drift/latest/drift.web/DriftWebStorage-class.html) implementation that delegates work to the web file system APIs and then use `WebDatabase.withStorage()` to open a web database that uses your file system storage implementation to persist the database.
